### PR TITLE
Made install a top level Meson command.

### PR DESCRIPTION
**Pull request by jpakkane**
*Not finished yet.*

Make installing a top level command that can be invoked with `meson install`. This allows us to add more functionality to the install command in the same way as we did to the test command. In a similar way you can tell the install command to not rebuild before running the install for those who really know what they are doing.

The primary use case is something like `meson install --only-changed` or something like that, which checks each file to be copied, and only does the copy if the target is either non-existing or is older than the file to be copied.

### CODE REVIEW

**Comment by xclaesse** [comment]
This will conflict badly with my command line rewrite PR. Do you think it would be possible to write this on top of my branch?

**Comment by jpakkane** [comment]
I will rebase it once that branch is in.

File: mesonbuild/mesonmain.py [hunk 1]
```diff
@@ -270,23 +266,32 @@ def run(original_args, mainfile):
...
             return rewriter.run(remaining_args)
         elif cmd_name == 'configure':
             try:
+                from . import mconf
                 return mconf.run(remaining_args)
             except MesonException as e:
                 mlog.exception(e)
                 sys.exit(1)
         elif cmd_name == 'wrap':
+            from .wrap import wraptool
```
Comment (ignatenkobrain) [inline mesonbuild/mesonmain.py]: __[Flake8]__

[F811] redefinition of unused 'wraptool' from line 26

File: mesonbuild/minstall.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+                if os.path.isdir(abs_dst):
+                    print('Tried to copy file %s but a directory of that name already exists.' % abs_dst)
+                if os.path.exists(abs_dst):
+                    os.unlink(abs_dst)
+                parent_dir = os.path.dirname(abs_dst)
+                if not os.path.isdir(parent_dir):
+                    os.mkdir(parent_dir)
+                    shutil.copystat(os.path.dirname(abs_src), parent_dir)
+                # FIXME: what about symlinks?
+                self.do_copyfile(abs_src, abs_dst)
```
Comment (nirbheek) [inline mesonbuild/minstall.py]: Why did this use `shutil.copy2()` earlier instead of `do_copyfile()`?

**Comment by jpakkane** [comment]
Now with rebasing and documentation. Please verify that I did not botch up the macOS install name fixes.

That is also why I'd want to merge this ASAP, because this has a renaming _and_ and an indentation change, there are a ton of merge conflicts all the time.

File: mesonbuild/scripts/meson_install.py [hunk 3]
```diff
@@ -1,441 +0,0 @@
...
-            if len(files) == 1:
-                return files[0]
-    return fname
-
-def install_targets(d):
-    for t in d.targets:
-        fname = check_for_stampfile(t.fname)
-        outdir = get_destdir_path(d, t.outdir)
-        outname = os.path.join(outdir, os.path.basename(fname))
-        final_path = os.path.join(d.prefix, t.outdir, os.path.basename(fname))
```
[the remark this answers is not in the document]
Comment (jpakkane) [reply mesonbuild/scripts/meson_install.py]: Thanks, should be fixed now.

**Pull request merged**